perception/autoware_ground_segmentation/src/scan_ground_filter/grid_ground_filter.cpp

```diff
@@ -0,0 +1,460 @@
+// Copyright 2024 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "grid_ground_filter.hpp"
+
+#include "data.hpp"
+
+#include <pcl/PointIndices.h>
+
+namespace autoware::ground_segmentation
+{
+
+void GridGroundFilter::convert()
+{
+  std::unique_ptr<ScopedTimeTrack> st_ptr;
+  if (time_keeper_) st_ptr = std::make_unique<ScopedTimeTrack>(__func__, *time_keeper_);
+
+  const size_t in_cloud_data_size = in_cloud_->data.size();
+  const size_t in_cloud_point_step = in_cloud_->point_step;
+
+  for (size_t data_index = 0; data_index + in_cloud_point_step <= in_cloud_data_size;
+       data_index += in_cloud_point_step) {
+    // Get Point
+    pcl::PointXYZ input_point;
+    data_accessor_.getPoint(in_cloud_, data_index, input_point);
+    grid_ptr_->addPoint(input_point.x, input_point.y, input_point.z, data_index);
+  }
+}
+
+void GridGroundFilter::preprocess()
+{
+  std::unique_ptr<ScopedTimeTrack> st_ptr;
+  if (time_keeper_) st_ptr = std::make_unique<ScopedTimeTrack>(__func__, *time_keeper_);
+
+  // eliminate empty cells from connection for efficiency
+  grid_ptr_->setGridConnections();
+}
+
+bool GridGroundFilter::recursiveSearch(
+  const int check_idx, const int search_cnt, std::vector<int> & idx) const
+{
+  // recursive search
+  if (check_idx < 0) {
+    return false;
+  }
+  if (search_cnt == 0) {
+    return true;
+  }
+  const auto & check_cell = grid_ptr_->getCell(check_idx);
+  if (check_cell.has_ground_) {
+    // the cell has ground, add the index to the list, and search previous cell
+    idx.push_back(check_idx);
+    return recursiveSearch(check_cell.scan_grid_root_idx_, search_cnt - 1, idx);
+  }
+  // if the cell does not have ground, search previous cell
+  return recursiveSearch(check_cell.scan_grid_root_idx_, search_cnt, idx);
+}
```

Comment:
[imho]

It will be safer to set `max_call_count` or something so that it prevents the infinite recursion. ( I think the end of recursion is within the closest grid to the ego, so it won't happen in this context.)